test/Tester/StorageFacet/Feature.Abstractions/IStorageFeature.cs

```diff
@@ -0,0 +1,58 @@
+﻿using System.Threading.Tasks;
+
+namespace Tester.StorageFacet.Abstractions
+{
+    /// <summary>
+    /// Primary storage feature interface.
+    ///  This is the actual functionality the users need.
+    /// </summary>
+    /// <typeparam name="TState"></typeparam>
+    public interface IStorageFeature<TState>
+    {
+        TState State { get; set; }
+
+        Task Save();
+
+        // Test calls - used to verify facet wiring works
+        string Name { get; }
+        string GetExtendedInfo();
+    }
+
+    /// <summary>
+    /// Feature configuration information which application layer can provide to the
+    ///  feature per instance (by grain type if using attributes).
+    /// </summary>
+    public interface IStorageFeatureConfig
+    {
+        string StateName { get; }
+    }
+
+    /// <summary>
+    /// Feature configuration utility class
+    /// </summary>
+    public class StorageFeatureConfig : IStorageFeatureConfig
+    {
+        public StorageFeatureConfig(string stateName)
+        {
+            this.StateName = stateName;
+        }
+
+        public string StateName { get; }
+    }
+
+    /// <summary>
+    /// Creates a storage feature from a configuration
+    /// </summary>
+    public interface IStorageFeatureFactory
+    {
+        IStorageFeature<TState> Create<TState>(IStorageFeatureConfig config);
```

Comment:
> - Having a typed factory would prohibit (or significantly complicate) managing IStorageFeatureFactorys as named services.
> - By having typed factories, each specialization would necessitate the construction of a new factory.

Is this right? we can still register `IStorageFeatureFactory<>` in the container (as an open generic), so that by default `IStorageFeatureFactory<Foo>` can be resolved without any additional registration.

> - Users that wished to have the factory injected in their grain constructor directly would need to have a factory for each type of data they wished to construct.

This is actually a plus, and supports the explicit dependency principle. It's the same reason why injecting `IServiceProvider` to get a specific service is worst than directly injecting that service (in application level code, of course, which is what we are talking about here). This also allows `IStorageFeatureFactory<Foo>`  and `IStorageFeatureFactory<Bar>` to use different factory implementations if they require so.

> Regarding the MakeGenericMethod, it is only called when creating the argument factory the first time, all subsequent calls simply invoke the method.

Isn't `StorageFeatureFactory.Create<TState>` calling `MakeGenericMethod` on every argument for every activation? I may be confused with the code flow though.